thunder/torch/__init__.py

```diff
@@ -4072,6 +4072,234 @@ def avg_pool3d(
+_torch_fallback_supported_module = [torch, torch.nn.functional, torch.Tensor, torch.special]
+
+
+def get_torch_fallback_operators_module(fn):
+    if fn in _ignored_torch_names:
+        return None
+    if fn in _torch_to_thunder_function_map:
+        return None
+    for m in _torch_fallback_supported_module:
+        if fn in _overridable_torch_ops[m]:
+            if hasattr(fn, "__name__") and fn.__name__.endswith("_"):
+                return None
+            return m
+    return None
+
+
+def _is_differentiable(arg):
+    from torch._subclasses.fake_tensor import FakeTensor
+
+    if isinstance(arg, (torch.Tensor, FakeTensor, TensorProxy)):
+        return dtypes.is_inexact_dtype(to_dtype(arg.dtype))
+    return False
+
+
+def _make_differentiable_wrapper(func, *args, **kwargs):
+    from thunder.core.pytree import tree_flatten, tree_unflatten
+
+    flat_args, spec = tree_flatten((args, kwargs))
+    differentiable_args_idx = tuple(i for i, a in enumerate(flat_args) if _is_differentiable(a))
+    differentiable_args = tuple(arg for i, arg in enumerate(flat_args) if i in differentiable_args_idx)
+
+    def wrapper(*diff_args):
+        diff_args_iter = iter(diff_args)
+        full_args = [next(diff_args_iter) if i in differentiable_args_idx else arg for i, arg in enumerate(flat_args)]
+        full_args, full_kwargs = tree_unflatten(full_args, spec)
+        return func(*full_args, **full_kwargs)
+
+    return wrapper, differentiable_args
+
+
+def register_torch_op(torchfn, fn_meta, m):
+    id_str = f"{m.__name__}.{torchfn.__name__}"
+    warnings.warn(f"Automatically register the {torchfn} in Thunder ({id_str})")
+    _fn = langctx(Languages.TORCH)(fn_meta)
+    sym = Symbol(
+        name=torchfn.__name__,
+        meta=_fn,
+        id=id_str,
+    )
+    _torch_to_thunder_function_map[torchfn] = sym
+    from thunder.core.interpreter import interpreter_needs_wrap
+    from thunder.core.jit_ext import (
+        _general_jit_lookaside_map,
+        ensure_recursive_proxies,
+        record_source_loc_in_symbol_header,
+    )
+
+    _general_jit_lookaside_map.update(
+        {torchfn: ensure_recursive_proxies(interpreter_needs_wrap(record_source_loc_in_symbol_header(sym)))}
+    )
+    from thunder.executors.torchex import _always_executable, ex
+
+    op = ex.register_operator(torchfn.__name__, module=m, meta=fn_meta)
+    ex.register_implementation(sym, op, checker=_always_executable)
+
+    from thunder.core.transforms import augmented_forward_impls, backward_impls
+
+    augmented_forward_impls[sym.id] = augmented_forward_adaptor(torchfn, op)
+
+    def _vjp_impl(residules, *gs) -> torch.Tensor:
+        assert isinstance(residules, dict)
+        inp_args = residules["inputs"][0]
+        inp_kwargs = residules["inputs"][1]
+        func = residules["torch_func"]
+
+        wrapped_func, diff_args = _make_differentiable_wrapper(func, *inp_args, **inp_kwargs)
```

Comment:
I was just wondering if we need to pass `func` from augmented forward to the `_vjp_impl` here. Instead can't `_vjp_impl` just capture the `torchfn` and use it?

Reply:
yes, I'll let it capture, thanks~